students/headieh_godwin/lesson09/donor_models.py

```diff
@@ -119,35 +115,115 @@ def write_thanks_all(self, name, total,numb):
+
+    def create_html_report(self):
+        """ opens tab with a report of donors"""
+        page = Html()
+        head = Head()
+        style = Style()
+        style.append(Customer('#customers {border-collapse: collapse;\
+         width: 50%;margin:auto;}'))
+        style.append(Customer('#customers th {text-align: center; \
+        color: c0c0c0; border-width: 1px 1px 4px 1px; border-style: solid;\
+         border-color:black; padding: 8px; padding: 12px, 0, 12px, 0;}'))
+        style.append(Customer('#customers tr:nth-child(even) {background-color:\
+         white}'))
+        style.append(Customer('#customers tr:nth-child(odd){background-color:\
+         lightgrey}'))
+        style.append(Customer('#customers td:nth-child(even) {text-align:\
+         right}'))
+        style.append(Customer('#customers td:nth-child(odd){text-align:\
+         center}'))
+        style.append(Customer('#customers td {border: 1px solid black;\
+        padding: 8px;}'))
+        style.append(Customer('body {background-image: \
+        linear-gradient(to bottom right, white ,black);\
+         background-repeat: no-repeat; width: 100vw; height: 100vh;}'))
+        head.append(style)
+        page.append(head)
+        body = Body()
+        body.append(H(1, 'Donation Report', style="font-style: oblique;\
+         color: 9c9c9c; text-align:center;"))
+        table = Table(id='customers')
+        column_list = Tr()
+        column_list.append(Th('Donor Name'))
+        column_list.append(Th('Total Given'))
+        column_list.append(Th('Num Gifts'))
+        column_list.append(Th('Average Gifts'))
+        table.append(column_list)
+
+        for this_ob in self.sort_donors():
+            donor_tr = Tr()
+            name = this_ob._name
+            num = this_ob.d_num
+            total = this_ob.d_tot
+            aveg = this_ob.d_avg
+            donor_tr.append(Td(name))
+            if this_ob.d_tot > 1000000:
+                donor_tr.append(Td(f"${total:<14.3e}"))
+            else:
+                donor_tr.append(Td(f"${total:<14.2f}"))
+            donor_tr.append(Td(f"{num:<15}"))
+
+            if this_ob.d_avg > 1000000:
+                donor_tr.append(Td(f"${aveg:<4.3e}"))
+            else:
+                donor_tr.append(Td(f"${aveg:<4.2f}"))
+            table.append(donor_tr)
+
+        body.append(table)
+        page.append(body)
+
+        parent = os.getcwd()
+        timestr = time.strftime("%Y%m%d-%H%M%S")
+        os.mkdir(timestr+'report')
+        filename = os.path.join(parent, timestr+'report' + '/donors_report.html')
+        f = io.StringIO()
+        page.render(f)
+        html_content = f.getvalue()
+        with open(filename, 'w') as outfile:
+            outfile.write(html_content)
+        if os.path.exists('/Applications/Safari.app'):
+            os.system(f'open /Applications/Safari.app {filename}')
+        else:
+            webbrowser.open(filename) #this did not work for me on safari
```

Comment:
highly recommend checking out https://pypi.python.org/pypi/tabulate and playing with it


it's a great library with different format options, I always suggest to use ready libraries vs trying to write code from scratch 🙂